### DISCUSSION TITLE
Issue with "super" script using IBM Notifier

### DISCUSSION THREAD

**Discussion by piwo41234** [IBM/mac-ibm-notifications#220] [General]
Hello

I have one user with weird issue 

IBM Notifier returns following output export PATH="/usr/local/opt/openjdk@17/bin:$PATH" 0  instead of 0 1 2 3 ... 

It happens on intel Mac with Sonoma. If possible i would like to ask for suggestion what  i should check or what may be causing it 

Regards
Krzysztof

**Comment by SMartorelli** [comment]
Hi 👋 could you please attach here the script you're using? Please remove any PI if needed. That would help me in understanding your issue.

**Comment by piwo41234** [comment]
Hello 

Thank you very much for a quick response . We are using very famous and very good super man script responsible for MacOS updates https://github.com/Macjutsu/super/blob/main/super

At the moment i have one user out of 1500 and i would like to know how to fix it because it may  reoccur.

Error appears in as output of thins this:  line dialog_response=$("${IBM_NOTIFIER_BINARY}" "${ibm_notifier_array[@]}") dialog_return="$?" 

Regards 
Krzysztof

**Comment by SMartorelli** [comment]
I think this is related to super and not to IBM Notifier. Seems like something else get caught instead of the IBM Notifier output. I would suggest you to interact with Kevin (@Macjutsu) to understand the source of the issue. I will convert this to a Discussion so that if needed we can continue the conversation there.

**Reply by piwo41234** [reply]
Hello

I already asked Kevin for any advice and he mentioned  that he is not using Java at all and it must be something related to IBM Notifier   :)

I know that its super hard to find what may be causing it and so i was hoping only for any advice and suggestions 

Regards
Krzysztof

**Reply by SMartorelli** [reply]
I'm afraid I don't have any useful suggestions for you then 😞 IBM Notifier doesn't make use of any java binary or dependency. It's unclear to me how one of those variables have been overridden by what it seems a PATH definition.

**Comment by Macjutsu** [comment]
So here is the bit of code used in `super` to run IBM Notifier.

```	
	# Start IBM Notifier and wait for ${dialog_response} and ${dialog_return}.
	unset dialog_response
	unset dialog_return
	dialog_response=$("${IBM_NOTIFIER_BINARY}" "${ibm_notifier_array[@]}")
	dialog_return="$?"
```

For clarification the `${ibm_notifier_array[@]}` should only contain IBM Notifier specific command flags and nothing more. @piwo41234  can you share the verbose super.log output that shows the content of that array? It's possible that `super` is putting something strange in that array.

**Comment by piwo41234** [comment]
Hello


Verbose logs attached just please search for jdk

Regards
Krzysztof

**Reply by Macjutsu** [reply]
Replying to a GitHub email does not include attachments.

Also, it's generally dangerous to ship detailed logs arround b/c there can be some sensitive identifying information in them.

Instead just search for the line with this text `ibm_notifier_array` just prior to the error.

**Comment by piwo41234** [comment]
Hi

For some reason  ibm_notifier_array is empty

Regards
Krzysztof

Fri Aug 23 12:16:16 L103858MUS super[98683]: Verbose Mode: Function
check_user_focus: user_focus_active is: FALSE
Fri Aug 23 12:16:16 L103858MUS super[98683]: Status: Hard maximum deferral
count of 3 NOT passed with 1 remaining.
Fri Aug 23 12:16:16 L103858MUS super[98683]: Verbose Mode: Function
check_deadlines_count: deadline_count_status is: FALSE
Fri Aug 23 12:16:16 L103858MUS super[98683]: Verbose Mode: Function
check_deadlines_count: user_focus_active is: FALSE
Fri Aug 23 12:16:16 L103858MUS super[98683]: Verbose Mode: Function
set_display_strings_language: display_string_workflow_title is: macOS
Sonoma 14.6 Minor Update
Fri Aug 23 12:16:16 L103858MUS super[98683]: IBM Notifier: Restart or defer
dialog with a 10 second timeout.
Fri Aug 23 12:16:16 L103858MUS super[98683]: Verbose Mode: Function
open_dialog_ibm_notifier: display_accessory_enabled is: TRUE
Fri Aug 23 12:16:17 L103858MUS super[98683]: Verbose Mode: Function
open_dialog_ibm_notifier: dialog_user_auth_enabled is:
Fri Aug 23 12:16:17 L103858MUS super[98683]: Warning: Unable to show
--dialog-timeout-* countdown due to the --display-accessory-* option.
However, there is still a display timeout at 10 seconds.
Fri Aug 23 12:16:17 L103858MUS super[98683]: Verbose Mode: Function
open_dialog_ibm_notifier: ibm_notifier_array is:
-type popup -bar_title macOS Sonoma 14.6 Minor Update Requires Restart
-icon_path /Library/Management/super/icon.png -icon_width 100 -icon_height
100 -secondary_button_label Restart -subtitle • 1 out of 3 deferrals
remaining.
 -main_button_label Defer -accessory_view_type image
-accessory_view_payload /opt/PegaLogo/Notification.png
-secondary_accessory_view_type dropdown -secondary_accessory_view_payload
/title Defer software update for: /list Defer 1 Hour
Defer 2 Hours
Defer 4 Hours
Defer 8 Hours
Defer 1 Day /selected 0 -timeout 10 -unmovable -position center
-always_on_top -disable_quit
Fri Aug 23 12:16:30 L103858MUS super[98683]: Verbose Mode: Function
open_dialog_ibm_notifier: dialog_response is: export
***@***.***/bin:$PATH"
4
Fri Aug 23 12:16:30 L103858MUS super[98683]: Verbose Mode: Function
open_dialog_ibm_notifier: dialog_return is: 0
Fri Aug 23 12:16:30 L103858MUS super[98683]: Verbose Mode: Function
exit_clean: Local preference file at clean exit:
/Library/Management/super/com.macjutsu.super:
{
    ASUListCache = 0;
    DeadlineCountHard = 3;
    DeadlineCounterHard = 2;
    DeferralTimerMenu = "60,120,240,480,1440";
    DialogTimeoutRestartOrDefer = 1800;
    DisplayAccessoryDefaultFile = "/opt/PegaLogo/Notification.png";
    DisplayAccessoryType = IMAGE;
    DisplayIconFileCachedOrigin = "/opt/PegaLogo/icon.png";
    DisplayIconSize = 100;
    DisplayUnmovable = ALWAYS;
    MacLastStartup = "2024-08-22:12:00:00";
    SuperStatus = "Fri Aug 23 12:16:16: Running: Dialog restart or defer.";
    SuperVersion = "4.0.3";
    TestMode = 1;
    VerboseMode = 1;
    WorkflowDisableRelaunch = 1;
    WorkflowZeroDateAutomatic = "2024-08-23:12:14";
    WorkflowZeroDateMacOSVersion = "14.6";
    macOSBetaProgramCache = 0;
    macOSNewestMarketingNameCache = 0;
    macOSNewestVersionCache = 0;
}
Fri Aug 23 12:16:30 L103858MUS super[98683]: **** S.U.P.E.R.M.A.N. 4.0.3 -
CLEAN EXIT ****
Fri Aug 23 12:17:03 L103858MUS super-starter[239]: **** S.U.P.E.R.M.A.N.
4.0.3 - LAUNCHDAEMON ****
Fri Aug 23 12:17:04 L103858MUS super[255]: **** S.U.P.E.R.M.A.N. 4.0.3 -
SUPER STARTUP WORKFLOW ****
Fri Aug 23 12:17:04 L103858MUS super[255]: Verbose Mode: Function
workflow_startup: Verbose mode enabled.
Fri Aug 23 12:17:04 L103858MUS super[255]: Verbose Mode: Function
workflow_startup: super_current_folder is: /Library/Management/super
Fri Aug 23 12:17:04 L103858MUS super[255]: Verbose Mode: Function
workflow_startup: Uptime is: 12:17  up 1 day, 16 mins, 3 users, load
averages: 3.79 4.85 4.87

**Comment by Macjutsu** [comment]
The array isn't empty.... it's literally all the lines following the log entry until the next log entry (each indivdual log entry starts with the date). However, as we can clearly see the error showing up in the:

```
Fri Aug 23 12:16:30 L103858MUS super[98683]: Verbose Mode: Function
open_dialog_ibm_notifier: dialog_response is: export
***@***.***/bin:$PATH"
4
```

**Comment by SMartorelli** [comment]
I don't think this is related to super or IBM Notifier. Just my two cents: to me, this seems like a one-time issue, possibly caused by the user's actions in the terminal while super was running. I can't really explain how that's possible.

**Comment by Macjutsu** [comment]
Good point, how are you running `super`? Pehaps you could kick it off using another method like a Jamf Pro Policy. In this case `super` will restart itself via LaunchDaemon and any weirdness with the user's shell enviornment should be ignored.

**Comment by piwo41234** [comment]
Hi Kevin

"In the meantime, I completely deleted Super from the Mac twice and
reinstalled it using our standard Jamf policy, which is working perfectly
fine. I also tried resetting Super multiple times using the
--reset-super command,
but without success.

Additionally, I found a JDK entry in the user’s /Users/user/.bash_profile that
appears in the Super log and removed it then restarted and tested one more
time  but it didn’t change anything.


Regards

Krzysztof

**Comment by Macjutsu** [comment]
And... you just confirmed it's due to customizations in the user space.

Try having `super` restart by itself by clearing the following preference and waiting a minute... also this will only log to the super.log

`sudo defaults delete /Library/Management/super/com.macjutsu.super.plist NextAutoLaunch`

**Comment by piwo41234** [comment]
Hi Kevin

I tested the command today with the user but unfortunately it didn't change
anything. User got 3 notifications in 10 min and the jdk entry is still in
the logs.

Regards
Krzysztof

**Comment by Macjutsu** [comment]
Find this bit of code in the script....

```
	0)
		if [[ -n "${deferral_timer_menu_minutes}" ]]; then
			deferral_timer_minutes="${deferral_timer_menu_minutes_array[${dialog_response}]}"
```

and add one line exactly like this modified example

```
	0)
		if [[ -n "${deferral_timer_menu_minutes}" ]]; then
			dialog_response="$(echo "${dialog_response}" | sed -e 's/[^0-9]*//g' | xargs)"
			deferral_timer_minutes="${deferral_timer_menu_minutes_array[${dialog_response}]}"
```

Lemme know how that works for you... probably adding this to the next beta.

**Comment by piwo41234** [comment]
Hi Kevin

I did what you suggested and we have small progress because at least now
plist has info about  NextAutoLaunch = "2024-09-09:14:41:00

However, there is still a weird entry in the logs   "* dialog_response is:
export ***@***.***/bin:$PATH"*  "

I checked *SuperStatus = "Mon Sep 09 14:41:46: Pending: User chose to defer
update for  minutes.";*   and as you can see information for how long user
postponed the update is missing

When user clicked to postpone update it was 14:41 and  NextAutoLaunch is
exactly the same 14:41 but should be 15:41

I don’t want to waste your valuable time, especially since you are working
on the new version

If you still want to troubleshoot I'm happy to follow your suggestions

Regards
Krzysztof

**Comment by Macjutsu** [comment]
I can't stop the weird text in `dialog_response`... the operating system is literally stickin that in there somehow. But somehow it's even worse in your last testing... now it's not even returning a single digit number a few lines later?

The proposed fix is trying to remove the extra text so `super` only "pays attention" to the number at the end... but it looks like it still didn't work because the number at the end is missing in the first place?

To be clear, I am unable to replicate this issue on my end, I can only speculate on how to resolve it.

**Comment by piwo41234** [comment]
Hi Kevin

Digit number is still there but i just didn't copy everything

Mon Sep 09 14:41:45 L103858MUS super[40368]: Verbose Mode: Function
open_dialog_ibm_notifier: dialog_response is: export
***@***.***/bin:$PATH"
2

Regards
Krzysztof

**Comment by Macjutsu** [comment]
What is the value of `dialog_return` from that last run?

**Comment by piwo41234** [comment]
Defer 1 Day /selected 0 -timeout 10 -unmovable -position center
-always_on_top -disable_quit
Mon Sep 09 14:41:45 L103858MUS super[40368]: Verbose Mode: Function
open_dialog_ibm_notifier: dialog_response is: export
***@***.***/bin:$PATH"
2
Mon Sep 09 14:41:46 L103858MUS super[40368]: Verbose Mode: Function
open_dialog_ibm_notifier: dialog_return is: 0
Mon Sep 09 14:41:46 L103858MUS super[40368]: Status: User chose to defer
update for  minutes.
Mon Sep 09 14:41:46 L103858MUS super[40368]: Verbose Mode: Function
set_auto_launch_deferral: deferral_timer_minutes is:
Mon Sep 09 14:41:46 L103858MUS super[40368]: Exit: super is scheduled to
automatically relaunch at: 2024-09-09:14:41:00
Mon Sep 09 14:41:46 L103858MUS super[40368]: Verbose Mode: Function
exit_clean: Local preference file at clean exit:

**Comment by Macjutsu** [comment]
Alrighty... try this now...

```
	0)
		if [[ -n "${deferral_timer_menu_minutes}" ]]; then
			dialog_response="$(echo "${dialog_response}" | sed -e 's/[^0-9]*//g' | tr -d '\n')"
			deferral_timer_minutes="${deferral_timer_menu_minutes_array[${dialog_response}]}"
```

**Comment by piwo41234** [comment]
Hi Kevin

Unfortunately it didn't change anything:

Tue Sep 10 13:36:09 L103858MUS super[37317]: Verbose Mode: Function
open_dialog_ibm_notifier: ibm_notifier_array is:
-type popup -bar_title macOS Sonoma 14.6 Minor Update Requires Restart
-icon_path /Library/Management/super/icon.png -icon_width 100 -icon_height
100 -secondary_button_label Restart -subtitle • 1 out of 3 deferrals
remaining.
 -main_button_label Defer -accessory_view_type image
-accessory_view_payload /opt/PegaLogo/Notification.png
-secondary_accessory_view_type dropdown -secondary_accessory_view_payload
/title Defer software update for: /list Defer 1 Hour
Defer 2 Hours
Defer 4 Hours
Defer 8 Hours
Defer 1 Day /selected 0 -timeout 10 -unmovable -position center
-always_on_top -disable_quit
Tue Sep 10 13:36:20 L103858MUS super[37317]: Verbose Mode: Function
open_dialog_ibm_notifier: dialog_response is: export
***@***.***/bin:$PATH"
1
Tue Sep 10 13:36:20 L103858MUS super[37317]: Verbose Mode: Function
open_dialog_ibm_notifier: dialog_return is: 0
Tue Sep 10 13:36:20 L103858MUS super[37317]: Status: User chose to defer
update for  minutes.
Tue Sep 10 13:36:20 L103858MUS super[37317]: Verbose Mode: Function
set_auto_launch_deferral: deferral_timer_minutes is:
Tue Sep 10 13:36:20 L103858MUS super[37317]: Exit: super is scheduled to
automatically relaunch at: 2024-09-10:13:36:00
Tue Sep 10 13:36:20 L103858MUS super[37317]: Verbose Mode: Function
exit_clean: Local preference file at clean exit:
/Library/Management/super/com.macjutsu.super:
{
    DeadlineCountHard = 3;
    DeadlineCounterHard = 2;
    DeferralTimerMenu = "60,120,240,480,1440";
    DialogTimeoutRestartOrDefer = 1800;
    DisplayAccessoryDefaultFile = "/opt/PegaLogo/Notification.png";
    DisplayAccessoryType = IMAGE;
    DisplayIconFileCachedOrigin = "/opt/PegaLogo/icon.png";
    DisplayIconSize = 100;
    DisplayUnmovable = ALWAYS;
    MacLastStartup = "2024-08-30:14:34:00";
    NextAutoLaunch = "2024-09-10:13:36:00";
    SuperStatus = "Tue Sep 10 13:36:20: Pending: User chose to defer update
for  minutes.";
    SuperVersion = "4.0.3";
    TestMode = 1;
    VerboseMode = 1;
    WorkflowDisableRelaunch = 1;
    WorkflowZeroDateAutomatic = "2024-09-10:13:34";
    WorkflowZeroDateMacOSVersion = "14.6";

**Comment by Macjutsu** [comment]
Give the latest beta a try please... https://github.com/Macjutsu/super/releases/tag/v5.0.0-beta3

Kevin

**Comment by piwo41234** [comment]
Hi Kevin

I have tested the new Beta 3, but nothing has changed. The entry is still
present in the logs.

The user did something unusual, but I can’t determine what exactly.

The user will receive a new Mac, which should definitely resolve this
issue.Thank you very much for all your help. I appreciate it very much

Regards
Krzysztof
